### PR TITLE
Weaken release sequences to match the C++20 memory model

### PR DESCRIPTION
**Pull request by JCTyblaidd**
See [Weaken Release Sequences](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0982r1.html), since the exception for relaxed writes on the same thread as a release write not blocking release sequences was removed in the C++20 memory model compared to the C11 memory model the paper was based on. The implementation can be updated and simplified to match this.  [Rust is currently specified to use the C++20 memory model](https://doc.rust-lang.org/std/sync/atomic/index.html).

### CODE REVIEW

File: src/data_race.rs [hunk 1]
```diff
@@ -185,13 +190,6 @@ struct AtomicMemoryCellClocks {
     /// happen-before a thread if an acquire-load is
     /// performed on the data.
     sync_vector: VClock,
-
-    /// The Hash-Map of all threads for which a release
-    /// sequence exists in the memory cell, required
-    /// since read-modify-write operations do not
-    /// invalidate existing release sequences.
-    /// See page 6 of linked paper.
-    release_sequences: VSmallClockMap,
```
Comment (RalfJung) [inline src/data_race.rs:194]: Wow this can be entirely removed? That's neat. :D

File: src/data_race.rs [hunk 2]
```diff
@@ -283,21 +282,20 @@ impl MemoryCellClocks {
...
     }
 
     /// Update the memory cell data-race tracking for atomic
     /// store relaxed semantics.
     fn store_relaxed(&mut self, clocks: &ThreadClockSet, index: VectorIdx) -> Result<(), DataRace> {
         self.atomic_write_detect(clocks, index)?;
+        
+        // The handling of release sequences was changed in C++20 and so
+        // the code here is different to the paper since now all relaxed
+        // stored block release sequences, the exception for same-thread
```
Comment (RalfJung) [inline src/data_race.rs:295]:
```suggestion
        // stores block release sequences. The exception for same-thread
```

**Comment by RalfJung** [comment]
Thanks, this looks great. :)

So the change only makes *fewer* programs defined, is that right? There is no program that had a data race before that is no longer considered to be racy with the new semantics?

**Comment by JCTyblaidd** [comment]
> So the change only makes _fewer_ programs defined, is that right? There is no program that had a data race before that is no longer considered to be racy with the new semantics?

Yes.

**Comment by RalfJung** [comment]
Great then, all ready to go!
@bors r+

**Pull request merged**